packages/perspective-jupyterlab/src/ts/renderer.ts

```diff
@@ -0,0 +1,249 @@
+import {ActivityMonitor} from '@jupyterlab/coreutils';
+import {ILayoutRestorer, JupyterFrontEnd, JupyterFrontEndPlugin} from '@jupyterlab/application';
+import {IThemeManager, WidgetTracker, Dialog, showDialog} from '@jupyterlab/apputils';
+import {ABCWidgetFactory, DocumentRegistry, IDocumentWidget, DocumentWidget} from '@jupyterlab/docregistry';
+import {PerspectiveWidget} from "@finos/perspective-phosphor";
+
+/**
+ * The name of the factories that creates widgets.
+ */
+const FACTORY_CSV = 'CSVPerspective';
+const FACTORY_JSON = 'JSONPerspective';
+
+const RENDER_TIMEOUT = 1000;
+
+type IPerspectiveDocumentType = "csv" | "json";
+
+// create here to reuse for exception handling
+const baddialog = () => {
+          showDialog({
+          body: "Perspective could not render the data",
+          buttons: [Dialog.okButton({ label: "Dismiss" })],
+          focusNodeSelector: "input",
+          title: "Error"});
+        };
+
+
+export class PerspectiveDocumentWidget extends DocumentWidget<PerspectiveWidget> {
+  constructor(options: DocumentWidget.IOptionsOptionalContent<PerspectiveWidget>, 
+              type: IPerspectiveDocumentType = "csv") {
+    super({ content: new PerspectiveWidget("test"), context: options.context, reveal: options.reveal});
+
+    this._psp = this.content;
+    this._type = type;
+    this._context = options.context;
+
+    this._context.ready.then(() => {
+      this._update();
+      this._monitor = new ActivityMonitor({
+        signal: this.context.model.contentChanged,
+        timeout: RENDER_TIMEOUT
+      });
+      this._monitor.activityStopped.connect(this._update, this);
+    });
+  }
+
+  private _update() {
+    try {
+      if(this._type === "csv") {
+        // load csv directly
+        const data: string = this._context.model.toString();
+        this._psp._update(data);
+      } else if (this._type === "json") {
+        const data = this._context.model.toJSON() as any;
+
+        if (Array.isArray(data) && data.length > 0) {
+          // already is records form, load directly
+          this._psp._update(data);
+
+        } else {
+          throw "Not handled";
+          // TODO
```

Comment:
Can we get rid of the comments/refactor it out to a separate function

Reply:
replaced with a comment describing the intent of what would go here, should anyone else feel like implementing it